rfcs/20180918-functions-not-sessions-20.md

```diff
@@ -0,0 +1,1123 @@
+# shape
+graph = f.graph_function(
+  (tf.float32, None)).graph
+
+f2 = tf.NewGraphFunction(
+  graph,
+  inputs=["Square:0"], 
+  outputs=["Square_1:0"])
+# The above may optionally take a
+# "prune" argument to allow for
+# pruning stateful operations in
+# `graph` that are not in the path
+# from inputs to outputs.
+f2(2.0) # 4.0</pre>
+
+
+   </td>
+  </tr>
+  <tr>
+   <td colspan="2" >Extending a graph
+   </td>
+  </tr>
+  <tr>
+   <td>
+
+
+
+<pre class="prettyprint">def build_graph():
+  x = tf.placeholder(tf.float32)
+  y = tf.square(x)
+  return y
+
+y = build_graph()
+z = tf.square(y)
+
+with tf.Session() as sess:
+  # Line below will return 16.0
+  sess.run(z, feed_dict={"Placeholder:0": 2.0))</pre>
+
+
+   </td>
+   <td>
+
+
+
+<pre class="prettyprint">@tf.function
+def f(x):
+  return tf.square(x)
+
+@tf.function
+def g(x):
+  return tf.square(f(x))
+
+g(2.0) # 16.0</pre>
+
+
+   </td>
+  </tr>
+</table>
+
+
+
+### Distributed Execution
+
+At the lowest level of the API, distributed execution continues to work with `tf.device` annotations, where the device name can reference remote devices as well, just like they do today.
+
+The `DistributionStrategy` API, typically aimed at synchronous training will continue to be the method of choice (where the API can be used inside a `function`). Other APIs such as go/tf-replicator will also be usable.
+
+The author realizes that this section can do with more detail. However, to keep this document more focused, these details will be discussed separately. In particular, usage of `MonitoredSession` and session hooks today needs additional thought.
+
+
+### `function`-ing Python control flow
+
+`function` decorates a graph construction function and transparently recreates graphs if needed. However, this does mean that if the function has data-dependent control flow then though the function will execute fine with eager execution enabled, `function` decorating it will fail. For example:
+
+
+```python
+def f(x, y):
+  if tf.equal(y, 0.0):
+    return y
+  return x / y
+
+x = tf.constant(2.0)
+y = tf.constant(2.0)
+
+f(x, y) # Will be 1.0
+
+df = tf.function(f)
+df(x, y)  # Will raise an error complaining about the data-dependent control flow
+```
+
+
+ \
+To fix this, one would have to use the graph construction APIs for control flow (`tf.cond`, `tf.while_loop`):
+
+
+```python
+def f(x, y):
+  return tf.cond(tf.equal(y, 0.0), lambda: y, lambda: x/y)
+
+x = tf.constant(2.0)
+y = tf.constant(2.0)
+
+f(x, y) # Will be 1.0
+
+df = tf.function(f)
+df(x, y)  # Will be 1.0
+```
+
+
+This situation can be improved with the help of  [autograph](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/python/autograph) to allow expression of control flow in Python. Whether autograph will be enabled by default or not is still under debate, but the option will be there as a flag on function. For example:
+
+
+```python
+df = tf.function(f, autograph=True)
+f(x, y) # Will be 1.0
+```
```

Comment:
FWIW, I think that doing `tf.function(autograph=True)(f)` would be a lot more natural, since that enables something like:

```python
@tf.function(autograph=True)
def f(x, y):
     ...
```

Reply:
Yup, fixed. Thanks!